Get etag from remote OC server

PVince81 opened the pull request:
Fixes https://github.com/owncloud/core/issues/15125

~~BUT this will break syncing with DAV servers that do not return any etags.~~
- ~~find a good solution for regular DAV servers, maybe differentiate between DAV and ownCloud backends (and use ownCloud backend for s2s too)~~ syncing should not break, it will fallback to use mtime for update detection

@icewind1991

Commit:
Get etag from remote OC server

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/79168345-193f-4b2a-ad72-2eee4a94ae82) was created.

Commit:
Added storage backend test for checkUpdate

This helps testing whether checkUpdate properly returns false when no
change exists

PVince81 commented:
- [x] add unit test

If you run `./autotest-external.sh sqlite webdav-ownCloud` it the test will fail on master but pass on this branch. If we had a "regular WebDAV" docker instance it would probably fail. (CC @MorrisJobke did you have any WIP for regular WebDAV docker, maybe lighttpd ?)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/b28cfe52-51ae-4383-80e8-a5ab7fc673f6) was created.

PVince81 commented:
- [ ] test against regular WebDAV server

PVince81 commented:
Raised https://github.com/owncloud/core/issues/15897 for the regular DAV docker.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/b28cfe52-51ae-4383-80e8-a5ab7fc673f6) completed: **1 new issues, 4 updated code elements**

PVince81 commented:
Revisiting this solution: this fixes the s2s etag detection issue (which doesn't always go through shareinfo.php).
In the case of regular WebDAV, the code will detect that no etag was returned and should rely on the "old" behavior and use mtime for change detection.

PVince81 commented:
To test:
- [ ] s2s: detect updates for single file shares, no repeated redownload
- [ ] s2s: detect updates for folder shares, no repeated redownload
- [ ] ownCloud ext storage mount: detect updates based on etag
- [ ] WebDAV ext storage mount: detect updates based on mtime

PVince81 commented:
Having trouble testing with lighttpd/webdav due to https://github.com/owncloud/core/issues/16681

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/73f54d67-74af-4edc-9fbc-9dc497524348) was created.

PVince81 commented:
Seems to work fine now. Let's see what the unit tests say.

DeepDiver1975 commented:
````
 Test\Files\Storage\Wrapper\Encryption.testCheckUpdate
 Stack Trace

Test\Files\Storage\Wrapper\Encryption::testCheckUpdate
Undefined index: storage_mtime

/var/jenkins/workspace/pull-request-analyser-ng-simple@4/label/SLAVE/lib/private/files/cache/watcher.php:89
/var/jenkins/workspace/pull-request-analyser-ng-simple@4/label/SLAVE/tests/lib/files/storage/storage.php:323

	0.051357	1
 Test\Files\Storage\Wrapper\Jail.testCheckUpdate
 Stack Trace

Test\Files\Storage\Wrapper\Jail::testCheckUpdate
No update
Failed asserting that true is false.

/var/jenkins/workspace/pull-request-analyser-ng-simple@4/label/SLAVE/tests/lib/files/storage/storage.php:324

	0.063577	1

PVince81 commented:
Seems to be `./autotest.sh sqlite tests/lib/files/storage/wrapper/encryption.php`
The mock isn't covering enough, I'll see if I can adapt it.

Commit:
Disable update check test when testing wrappers

The checkUpdate test is intended for testing real storages, not
wrappers.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/16642b86-3da2-4157-aa2a-43a9ad810b49) was created.

PVince81 commented:
- [x] TEST: unit tests pass for ownCloud external storage
- [x] TEST: unit tests pass for Webdav external storage: (or at least they're not breaking more than they were)

MorrisJobke commented:
@icewind1991 @nickvergessen mind to review this?

icewind1991 commented:
:+1: looks good

nickvergessen commented:
:+1: fixes the issue I had when writting a smashbox test for remote shares

nickvergessen commented:
@PVince81 8.0 suffers from the same issue I guess?

PVince81 commented:
Very likely.

@karlitschek backport to stable8 ?

PVince81 commented:
PR here: https://github.com/owncloud/core/pull/16734
